PYIC-8790: Add progress spinner

AmritSidhu opened the pull request:
## Proposed changes
### What changed

- Replace 'Continue...' button with progress spinner component on 'page-ipv-success' and 'page-ipv-reuse' pages
- Default content and translations are already built in to the component. The component is actually a link masquerading as a button (it takes an href) so we’ll need to use our journey GET mechanism to simulate a ‘next’ event.
- gate progress button behind feature flag

### Why did it change

- To ensure we can handle any potential latency / wait time gracefully this story is to replace the current Continue to service button on those two pages with a progress button component that shows a dynamic ‘waiting’ progress element.

### Issue tracking
<!-- Jira ticket & other docs, like RFCs -->

- [PYIC-8790](https://govukverify.atlassian.net/browse/PYIC-8790)

## Checklists

- [ ] READMEs and documentation up-to-date
- [ ] Browser/ unit/ Selenium tests have been written/ updated
- [ ] No risk of exposure: PII, credentials, etc through logs/ code
- [ ] Ensure added/updated routes have CSRF protection if required


[PYIC-8790]: https://govukverify.atlassian.net/browse/PYIC-8790?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-8790: Add progress spinner

To ensure we can handle any potential latency / wait time gracefully this story is to replace the current Continue to service button on those two pages with a progress button component that shows a dynamic ‘waiting’ progress element.

blakeyp reviewed (dismissed):
Looks good but hold merge until QA complete in dev

AmritSidhu committed:
PYIC-8790: Update frontend-ui package

AmritSidhu committed:
PYIC-8790: Rerun browser tests as part of upgrading frontend-ui

AmritSidhu committed:
PYIC-8790: Fix error page path

AmritSidhu committed:
PYIC-8790: add static unrecoverable PYI technical error route to bypass journey validation and support frontend-ui timeout handling

AmritSidhu committed:
PYIC-8790: use frontend-ui progress button macro with dev timeout override

AmritSidhu committed:
PYIC-8790: gate progress button timeout behind feature flag

- Add ENABLE_PROGRESS_SPINNER_BUTTON_TIMEOUT + PROGRESS_SPINNER_BUTTON_TIMEOUT config/env vars
- Expose values via res.locals for Nunjucks templates
- Conditionally pass customDevErrorTimeout to frontendUiProgressButton
- Wire env vars through ECS task definition and environment mappings

AmritSidhu committed:
PYIC-8790: Update snapshot tests

The update to the frontend-ui package is casuing a potential pixel shift

AmritSidhu committed:
PYIC-8790: Progress should be set to false in integration while testing

AmritSidhu committed:
PYIC-8790: gate progress button and initialisation behind feature flag

featureFlag entire button rather than the timeout config so that its safer while we test up to staging

thebauSoftwire commented on deploy/template.yaml:
```diff
@@ -228,6 +236,7 @@ Mappings:
       spinnerRequestLongWaitInterval: 60000
       mamSpinnerRequestTimeout: 600000
       dadSpinnerRequestTimeout: 1800000
+      enableProgressSpinnerButton: false
```
is there a reason we don't want a progressSpinnerButtonTimeout in int/prod?

thebauSoftwire commented on views/ipv/page/page-ipv-reuse.njk:
```diff
@@ -65,5 +66,32 @@
-    {{ journeyNextForm(pageId, csrfToken, 'general.buttons.continueToService' | translate ) }}
+    {% if enableProgressSpinnerButton %}
+      {# New progress button (experiment) #}
+      {% set customTimeout = progressSpinnerButtonTimeout %}
```
out of curiosity, if we don't set a value of `progressSpinnerButtonTimeout`, what value does `customTimeout` take and does the frontendUiProgressButton component handle it?

AmritSidhu replied:
@thebauSoftwire If the value is undefined (because it hasn’t been set), it defaults to the normal timing of 5 seconds, then transitions and waits another 5 seconds. This isn’t long enough in lower environments, which is why there was pushback to add support for a custom timeout.

Because this is currently an untested feature, Patrick asked that it only be enabled up to staging. For now, int/prod will continue to use the new button while we test through staging.

There will be a follow-up ticket to clean this up and enable the functionality for int/prod. Hopefully that answers your above comment too 🙂

thebauSoftwire replied:
thanks for clarifying :)

blakeyp commented on package.json:
```diff
@@ -76,14 +76,14 @@
     "@govuk-one-login/frontend-language-toggle": "2.2.1",
     "@govuk-one-login/frontend-passthrough-headers": "1.3.2",
-    "@govuk-one-login/frontend-ui": "4.0.1",
+    "@govuk-one-login/frontend-ui": "4.2.2",
```
I assume the updates to all the snapshots are just an artefact of bumping this package?

AmritSidhu replied:
Yes this is a result of the package bump

blakeyp commented on views/ipv/page/page-ipv-reuse.njk:
```diff
@@ -65,5 +66,32 @@
+    {% else %}
+      {# Original button (control) #}
+      {{ journeyNextForm(pageId, csrfToken, 'general.buttons.continueToService' | translate ) }}
+    {% endif %}
```
Before merging it's probably worth testing both a dev deployment with enableProgressSpinnerButton=true and a dev deployment with enableProgressSpinnerButton=false (for regression) ?

AmritSidhu replied:
- I have tested true/false and undefined in my dev env for enableProgressSpinnerButton
- I have also tested different times and undefined in my dev env for progressSpinnerButtonTimeout
